### PR TITLE
ENH: stats can use np.random.Generator

### LINKED ISSUE #11398: Update stats to cope with `np.random.Generator` machinery
`np.random.RandomState` is now regarded as legacy for the purposes of generating random variates. `np.random.Generator` is its replacement. We should upgrade the `scipy.stats` machinery to be able to use the latter.
For the most part this will involve searching the `scipy.stats` code for uses of things like `_random_state.random_sample` which exists in `random.RandomState`, but not in `random.Generator`. For that particular example it's replacement would be `_random_state.random`. They're fairly easy replacements. The harder part might be updating docstrings, and the test machinery.

### PR DESCRIPTION
**Pull request by andyfaff**
closes #11398. Allows `scipy.stats` to use `np.random.Generator` as well as the older legacy `np.random.RandomState`.

### CODE REVIEW

**Comment by chrisb83** [comment]
Do we need to edit check_random_state in scipy._lib._util as well?

**Comment by andyfaff** [comment]
Already been done by someone else. The test changes use a Generator.

File: scipy/stats/_discrete_distns.py
Comment (rlucas7) [inline scipy/stats/_discrete_distns.py]: With the change in numpy random is it worthwhile to add a convenience function over with`check_random_state()` in `_lib._util` ? Not a blocker though.

File: scipy/stats/_distn_infrastructure.py
Comment (rlucas7) [inline scipy/stats/_distn_infrastructure.py]: I realize this is not intended to change anything numerically and (presumably) the outputs are unchanged. Are time/benchmarks impacted at all? I'd expect not to be and not shouldn't be changed in this PR if impacted but wanted to call out to make sure we know whether any change.

File: scipy/stats/_distn_infrastructure.py
Comment (rlucas7) [inline scipy/stats/_distn_infrastructure.py]: Do we want to add anything in docstrings anywhere about the numpy version needed for the Generator? I have no strong feelins on yes/no. I'm not sure what the typical user would know in this case.

File: scipy/stats/tests/test_binned_statistic.py
Comment (rlucas7) [inline scipy/stats/tests/test_binned_statistic.py]: does this preserve the same rng stream? or might the randomness be different depending on the numpy in the env? I see that the seed is the same, I'm not familiar w/the spec(s) on new numpy random and if the generator would produce same stream or not. Not a blocker provided all tests are passing (they do for me locally) but wanted to clarify.

File: scipy/stats/tests/test_kdeoth.py
Comment (rlucas7) [inline scipy/stats/tests/test_kdeoth.py]: same comment as up in `common_tests.py`, this cannot  fail the test, not clear to me what is the intent on these lines of code.

File: scipy/stats/tests/common_tests.py
Comment (rlucas7) [inline scipy/stats/tests/common_tests.py]: wouldn't we want this to be an xfail type of situation that we would mark based on numpy version? Maybe I'm misunderstanding the intent of the code but it doesn't look like it could ever fail, am I missing something?

**Review by rlucas7 — COMMENTED** [review]
@andyfaff these are some nice refactors, I'm less familiar with the new numpy random setup so pardon the questions. Mostly clarifications for personal edification in my comments, a couple non-blockers, and 1 or 2 things I'm not clear on.

**Comment by rlucas7** [comment]
Not sure how far back we are supporting numpy but in my virtual env I still have an ***old*** numpy version and these changes still build for me locally (presumably that is expected and good to confirm). 

```
>>> import numpy 
>>> numpy.__version__
'1.16.4'
>>> import scipy 
>>> scipy.__version__
'1.5.0.dev0+997b4b4'
```
One of these days I'll get around to updating my virtual environment's numpy version.
All tests pass for me locally on my macOS.

**Comment by andyfaff** [comment]
The tests will pass on old numpy's because they're guarded by:

```
try:
    generator = np.random.default_rng()
except AttributeError:
    # older numpy can't run tests using the new Generators
    return
```

File: scipy/stats/_discrete_distns.py
Comment (andyfaff) [reply scipy/stats/_discrete_distns.py]: Unfortunately `Generator` and `RandomState` don't have a method with the same name that produces integers. There was only one usage so I didn't write a shim class.

File: scipy/stats/_distn_infrastructure.py
Comment (andyfaff) [reply scipy/stats/_distn_infrastructure.py]: `RandomState.{uniform, random_sample}` give the same stream of floats if setup with the same seed, so existing behaviour is unchanged. I changed the usage to `uniform` because both `Generator` and `RandomState` have a `uniform` method.

```
>>> from scipy._lib._util import check_random_state
>>> rng = check_random_state(9845)
>>> print(rng.uniform(size=5))
[0.0140971  0.71189422 0.70789811 0.9087384  0.23947253]
>>> rng = check_random_state(9845)
>>> print(rng.random_sample(size=5))
[0.0140971  0.71189422 0.70789811 0.9087384  0.23947253]
```

Here are some timings:
```
>>> from numpy.random import RandomState
>>> rs = RandomState(9845)

>>> %timeit rs.uniform(size=1000)
13.8 µs ± 314 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
>>> %timeit rs.uniform(size=1)
3.15 µs ± 72.3 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

>>> %timeit rs.random_sample(size=1000)
10.3 µs ± 394 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
>>> %timeit rs.random_sample(size=1)
1.26 µs ± 60 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

>>> gen = np.random.default_rng()
>>> %timeit gen.uniform(size=1000)
8.42 µs ± 135 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
>>> %timeit gen.uniform(size=1)
3.27 µs ± 98.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Only requesting a single float will incur a (small) performance hit. When one asks for lots of floats then a `Generator` is faster. The `RandomState.uniform` method is not as fast as `RandomState.random_sample`.

File: scipy/stats/_distn_infrastructure.py
Comment (andyfaff) [reply scipy/stats/_distn_infrastructure.py]: The user will probably realise quite quickly that it won't work if they can't create a Generator with their own numpy. I don't know where that information would go though, there are a lot of places that use random numbers.

File: scipy/stats/tests/test_binned_statistic.py
Comment (andyfaff) [reply scipy/stats/tests/test_binned_statistic.py]: It will generate the same rng stream. The previous code set the global state of the numpy random number generator, whereas these changes create a `RandomState` instance that is used solely for this test class. This means that the test should use exactly the same set of random numbers everytime because it eliminates the possibility that other tests (possibly run in parallel) can affect the state of the global generator.

File: scipy/stats/tests/common_tests.py
Comment (andyfaff) [reply scipy/stats/tests/common_tests.py]: `np.random.Generator` only came in 1.17, so the test shouldn't be run if numpy is too old, but we want it to run if it's available. I don't think marking the whole test as xfail (for numpy version) is right here, because it would make the whole test redundant on older numpy. So this specific test would have to be moved into its own function and marked as xfail. However, I don't think that can be done either because `check_random_state_property` is not a  pytest, it's only used by a parent pytest.

File: scipy/stats/tests/test_kdeoth.py
Comment (andyfaff) [reply scipy/stats/tests/test_kdeoth.py]: The purpose of the test is to perform a smoke test of whether a `Generator` will work when given to `gkde_trail.resample`. If Generator isn't available (old numpy) then the test is skipped, but if it is available then it should be run. Perhaps a better way is:
```
hasattr(np.random, 'Generator'):
    rng = np.random.default_rng(1234)
    gkde_trail.resample(n_sample, seed=rng)
```

File: scipy/stats/tests/test_binned_statistic.py
Comment (andyfaff) [reply scipy/stats/tests/test_binned_statistic.py]: BTW the random number streams from Generator and RandomState are different, even if the same seed is used.

File: scipy/stats/_discrete_distns.py
Comment (rlucas7) [reply scipy/stats/_discrete_distns.py]: it looks like it only shows up in 1 spot in the PR so not a blocker, if we start to see this across PRs we may want to reconsider.

File: scipy/stats/_distn_infrastructure.py
Comment (rlucas7) [reply scipy/stats/_distn_infrastructure.py]: cool, I'd imagine the single float request is less common than several so this may even improve speed w/new `Generator` in some cases-great. Thanks for putting the timings together.

File: scipy/stats/_distn_infrastructure.py
Comment (rlucas7) [reply scipy/stats/_distn_infrastructure.py]: yeah, and because it is so far across things, the release notes are probably the best place for this sort of thing anyway, so no need for changes in dcostring on this item.

File: scipy/stats/tests/test_binned_statistic.py
Comment (rlucas7) [reply scipy/stats/tests/test_binned_statistic.py]: ah I think I got it, stream will change because seed/state setting goes from global to local, but two streams w/same seed should generate the same seed. I'm mostly wondering in case we start seeing some spurious test failures on travis w.r.t. tests in `stats` using random numbers. For now it seems ok so no changes needed here. Thanks for clarifying.

Comment (andyfaff) [reply scipy/stats/tests/test_binned_statistic.py]: When I was running the test suite locally this particular test would fail about half the time, even though I hadn't touched it's code.

Comment (rlucas7) [reply scipy/stats/tests/test_binned_statistic.py]: ah ok, just curious if it has anything to do with the standard deviation calculatio/stability test? IIRC I set the exponent, `+ 1e6` so that only bits around 4 to 6 decimals in would lose stability, with a different stream that might change. If that was the test that was failing you could bump to `+ 1e7` and might make the numeric error more stable. If it wasn't that one then we'd need to look into it further if still failing on travis.

File: scipy/stats/tests/common_tests.py
Comment (rlucas7) [reply scipy/stats/tests/common_tests.py]: ah, I think I understand now. The rng streams differ so we cannot have a test here that the values are equal or evan an almost equal. Is that right? 
I like the `hasattr()` approach, between that and the comment in line above about numpy version it will be more clear what is happening. 
I'm still stuck on an assertion here that could fail, but we'd want it only to fail on environments w/newer numpy.
The scope of this PR it seems is really upgrading numpy random api inside SciPy and not tests per se. 
I also don't have any idea here and I don't want to hold up a perfectly good improvement on this issue if I can't offer a solution. Hopefully someone can come up with a way to do the test in a future PR. So let's call this comment non-blocking.

File: scipy/stats/tests/test_kdeoth.py
Comment (rlucas7) [reply scipy/stats/tests/test_kdeoth.py]: yeah see comment above, let's call this a non-blocking comment and hopefully someone finds a way to add a test in future PR.

**Review by rlucas7 — APPROVED** [review]
@andyfaff I've approved these changes look good and will update the stats tests to use the new API. I assume there will be more of these numpy random api change PRs coming (if not already in queue).
Thanks for clarifying on the rng stream and the timings. 
EDIT: I didn't see notice that the builds were still running from the PR, let's let those finish and I'll take a look tomorrow to see if anything fails and if anything needs to change.

**Comment by rlucas7** [comment]
It looks like the Travis build is the know fail ppc so no issues there.
There are some docstring changes in Thai pr so the other fails may be related but cannot tell. @andyfaff Are the doc build fails related to the changes or spurious or known fails? I cannot really tell from looking at the logs on my phone, the circle Ci UI for phone is really poor.

**Comment by rlucas7** [comment]
> It looks like the Travis build is the know fail ppc so no issues there. There are some docstring changes in Thai pr so the other fails may be related but cannot tell. @andyfaff Are the doc build fails related to the changes or spurious or known fails? I cannot really tell from looking at the logs on my phone, the circle Ci UI for phone is really poor.
> […](#)
> On Jan 24, 2020, at 9:56 PM, Andrew Nelson ***@***.***> wrote: ﻿ @andyfaff commented on this pull request. In scipy/stats/tests/test_binned_statistic.py: > @@ -13,13 +14,13 @@ class TestBinnedStatistic(object): @classmethod def setup_class(cls): - np.random.seed(9865) - cls.x = np.random.random(100) - cls.y = np.random.random(100) - cls.v = np.random.random(100) - cls.X = np.random.random((100, 3)) - cls.w = np.random.random(100) - cls.u = np.random.random(100) + 1e6 + rng = check_random_state(9865) When I was running the test suite locally this particular test would fail about half the time, even though I hadn't touched it's code. — You are receiving this because you commented. Reply to this email directly, view it on GitHub, or unsubscribe.

I got a chance to look at this on laptop, it looks like the html builds in the makefile are failing the subprocess call(s). I'm not sure if this has anything to do w/the docstring changes in this PR or is somewhere else. The log file(s) are very cryptic with useful information on this one.

**Comment by andyfaff** [comment]
It's a bit cryptic, but I think the reason the build is failing is:
`/home/circleci/repo/build/testenv/lib/python3.7/site-packages/scipy/stats/__init__.py:docstring of scipy.stats.exponweib:19: WARNING: py:obj reference target not found: numpy.random.RandomState.weibull`
This PR should work as soon as #11396 is merged (not an onerous review to do), but will require a few merge fixes.

**Comment by andyfaff** [comment]
I rebased which seemed to require a force push

**Comment by rlucas7** [comment]
> I rebased which seemed to require a force push

Looks like everything (except ppc) is passing the builds. I've approved the PR already but the builds were still running, this PR LGTM. Hopefully someone else can take a look and merge this PR after  #11396 goes into master. If no one gives a review I will hopefully have some time next weekend to review  #11396, @andyfaff  feel free to ping me on that PR if no comments on Friday.

**Comment by andyfaff** [comment]
#11396 was merged last week

**Review by rlucas7 — APPROVED** [review]
Forgot to do so before, marking as approving the changes.

**Comment by rlucas7** [comment]
I must of forgot to refresh my browser tab that was showing the PR open.

**Comment by andyfaff** [comment]
ping @rlucas7 for merge?

**Comment by rlucas7** [comment]
@andyfaff The PR looks good to me and looks ready to merge-these changes are fairly straightforward. Unfortunately I don't have merge rights to the repo so someone else will need to merge.

**Comment by ev-br** [comment]
LGTM too. Thanks Andrew, Lucas, Christoph. Merging.

This should probably get a short release note, too.

**Comment by andyfaff** [comment]
@ev-br release note - done.

**Pull request merged**
